common/flatpak-dir.c

```diff
@@ -14899,72 +14944,48 @@ flatpak_dir_list_unused_refs_with_options (FlatpakDir         *self,
   if (!flatpak_dir_is_user (self))
     {
       g_autoptr(GFile) user_base_dir = flatpak_get_user_base_dir_location ();
-
       if (g_file_query_exists (user_base_dir, cancellable))
         {
           g_autoptr(FlatpakDir) user_dir = flatpak_dir_get_user ();
-          g_auto(GStrv) user_app_refs = NULL;
-          g_auto(GStrv) user_runtime_refs = NULL;
-          g_autoptr(GHashTable) user_runtimes = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
 
-          if (!flatpak_dir_list_refs (user_dir, "app", &user_app_refs, cancellable, error))
+          if (!find_used_refs (self, user_dir, arch, metadata_injection, excluded_refs_ht,
+                               used_refs, cancellable, error))
             return NULL;
-
-          /* metadata_injection is NULL because it's not for this installation */
-          find_used_refs_for_apps (user_dir, self, user_app_refs, arch, NULL, used_runtimes, used_refs);
-
-          if (!flatpak_dir_list_refs (user_dir, "runtime", &user_runtime_refs, cancellable, error))
-            return NULL;
-
-          for (i = 0; user_runtime_refs[i] != NULL; i++)
-            {
-              const char *ref = user_runtime_refs[i];
-              g_assert (g_str_has_prefix (ref, "runtime/"));
-              g_hash_table_add (user_runtimes, (char *)ref + strlen ("runtime/"));
-            }
-
-          find_used_refs_for_runtimes (user_dir, self, NULL, user_runtimes, used_refs);
         }
     }
 
-  find_used_refs_for_runtimes (self, NULL, metadata_injection, used_runtimes, used_refs);
+  if (!flatpak_dir_list_refs (self, "runtime", &runtime_refs, cancellable, error))
+    return NULL;
+
+  refs = g_ptr_array_new_with_free_func (g_free);
 
   for (i = 0; runtime_refs[i] != NULL; i++)
     {
       const char *ref = runtime_refs[i];
-      g_auto(GStrv) parts = g_strsplit (ref, "/", -1);
 
-      if (arch != NULL && strcmp (parts[2], arch) != 0)
+      if (g_hash_table_contains (used_refs, ref))
         continue;
 
-      if (flatpak_dir_ref_is_pinned (self, ref))
-        {
-          g_debug ("Ref %s is pinned, considering as used", ref);
-          continue;
-        }
+      if (!ref_is_arch (ref, arch))
+        continue;
 
-      if (!g_hash_table_contains (used_refs, ref) &&
-          g_hash_table_add (refs_hash, (gpointer) ref))
+      if (filter_by_eol)
         {
-          if (!filter_by_eol)
-            g_ptr_array_add (refs, g_strdup (ref));
-          else
+          g_autoptr(GBytes) deploy_data = NULL;
+          deploy_data = flatpak_dir_get_deploy_data (self, ref, FLATPAK_DEPLOY_VERSION_ANY,
+                                                     cancellable, error);
+          if (deploy_data == NULL)
+            return NULL;
+
+          if (flatpak_deploy_data_get_eol (deploy_data) == NULL &&
```

Comment:
I think the filter-by-eol feature also needs to take the injected metadata into consideration. It may be that the updated version is eol, but the deployed is not.

Reply:
Ugh, eol data is from the commit, not the metadata...